### ISSUE TITLE
config_read_file() fails to parse an array of groups

### ISSUE THREAD

**Issue by voidtrance** [hyperrealm/libconfig#79]
I've run into something that I believe is a bug - config_read_file() fails to parse an array of groups.

I have the following example program that illustrates the issue:
```#include <libconfig.h>
 #include <stdio.h>
 #include <sys/stat.h>
 #include <errno.h>
 
 int create_config(config_t *cfg)
 {
	config_setting_t *setting, *array, *group, *elem, *root;

	root = config_root_setting(cfg);
	setting = config_setting_add(root, "test1", CONFIG_TYPE_INT);
	config_setting_set_int(setting, 2);
	array = config_setting_add(root, "array", CONFIG_TYPE_ARRAY);
	group = config_setting_add(array, "array", CONFIG_TYPE_GROUP);
	elem = config_setting_add(group, "elem1", CONFIG_TYPE_INT);
	config_setting_set_int(elem, 2);
	elem = config_setting_add(group, "elem2", CONFIG_TYPE_STRING);
	config_setting_set_string(elem, "element string");

	setting = config_setting_add(root, "string2", CONFIG_TYPE_STRING);
	config_setting_set_string(setting, "another string test");

	return 0;
}

int read_config(config_t *cfg)
{
	config_setting_t *setting, *array, *group, *elem, *root;
	const char *str;
	int i;

	root = config_root_setting(cfg);
	if (config_setting_lookup_string(root, "string2", &str))
		printf("string2: %s\n", str);
	else
		return 1;

	return 0;
}

int main(int argc, char **argv)
{
	config_t cfg;
	const char *str, filename[] = "test.cfg";
	struct stat st;
	int ret;

	config_init(&cfg);
	if (stat(filename, &st) == -1 && errno == ENOENT) {
		printf("config file does not exist\n");
		create_config(&cfg);
	} else {
		if (!config_read_file(&cfg, "test.cfg")) {
			printf("failed to read config file: %s:%u - %s\n",
			       config_error_file(&cfg), config_error_line(&cfg),
			       config_error_text(&cfg));
			goto bail;
		}
		else
			if (read_config(&cfg))
				printf("error looking up value\n");
	}

	if (!config_write_file(&cfg, "test.cfg"))
		fprintf(stderr, "Error writing config file\n");
bail:
	config_destroy(&cfg);
	return 0;
}
```
To compile use:
`gcc -g -O0 config.c -o config -lconfig`

When run for the first time, the program will generate the following config file:
```
test1 = 2;
array = [ 
  {
    elem1 = 2;
    elem2 = "element string";
  } ];
string2 = "another string test";
```
When run a second time, the program will attempt to read in the generated config file and will generate the following error:
`failed to read config file: test.cfg:3 - syntax error`

The same program works correctly if the type of _array_ is changed from CONFIG_TYPE_ARRAY to CONFIG_TYPE_LIST.

Normally, I would attribute this to me using the incorrect type for _array_. My initial though was that the CONFIG_TYPE_ARRAY type does not support groups as children. However, the documentation in the repository's _doc_ directory says that adding a child to an array, group, or list will return NULL if the type is invalid. This obviously does not happen.

On second thought, this may be a bug with config_setting_add() rather than config_read_file()

**Comment by jltallon** [comment]
On 02/08/17 18:08, voidtrance wrote:
>
> I've run into something that I believe is a bug - config_read_file() 
> fails to parse an array of groups.
>


Will take a look at this, as it is quite probably a valid bug.
I can't commit time to this before late next week, though.

Thank you for the complete report!

> I have the following example program that illustrates the issue:
>
> |#include <stdio.h> #include <sys/stat.h> #include <errno.h> int 
> create_config(config_t *cfg) { config_setting_t *setting, *array, 
> *group, *elem, *root; root = config_root_setting(cfg); setting = 
> config_setting_add(root, "test1", CONFIG_TYPE_INT); 
> config_setting_set_int(setting, 2); array = config_setting_add(root, 
> "array", CONFIG_TYPE_ARRAY); group = config_setting_add(array, 
> "array", CONFIG_TYPE_GROUP); elem = config_setting_add(group, "elem1", 
> CONFIG_TYPE_INT); config_setting_set_int(elem, 2); elem = 
> config_setting_add(group, "elem2", CONFIG_TYPE_STRING); 
> config_setting_set_string(elem, "element string"); setting = 
> config_setting_add(root, "string2", CONFIG_TYPE_STRING); 
> config_setting_set_string(setting, "another string test"); return 0; } 
> int read_config(config_t *cfg) { config_setting_t *setting, *array, 
> *group, *elem, *root; const char *str; int i; root = 
> config_root_setting(cfg); if (config_setting_lookup_string(root, 
> "string2", &str)) printf("string2: %s\n", str); else return 1; return 
> 0; } int main(int argc, char **argv) { config_t cfg; const char *str, 
> filename[] = "test.cfg"; struct stat st; int ret; config_init(&cfg); 
> if (stat(filename, &st) == -1 && errno == ENOENT) { printf("config 
> file does not exist\n"); create_config(&cfg); } else { if 
> (!config_read_file(&cfg, "test.cfg")) { printf("failed to read config 
> file: %s:%u - %s\n", config_error_file(&cfg), config_error_line(&cfg), 
> config_error_text(&cfg)); goto bail; } else if (read_config(&cfg)) 
> printf("error looking up value\n"); } if (!config_write_file(&cfg, 
> "test.cfg")) fprintf(stderr, "Error writing config file\n"); bail: 
> config_destroy(&cfg); return 0; } |

**Comment by hyperrealm** [comment]
Arrays don't support groups. They only support scalar types.

**Comment by voidtrance** [comment]
The fact that arrays don't support groups was my first guess of why this didn't work. However, this confirms the fact that the bug is in config_setting_add() instead of config_read_file().

**Comment by voidtrance** [comment]
I've looked through the code for config_setting_add(), config_setting_create(), and __config_list_add() and there is nothing that prevents the addition of a non-scalar values to a setting of type CONFIG_TYPE_ARRAY.

@jltallon, since you said that you are busy, I've created the following patch to address this:

```
diff --git a/lib/libconfig.c b/lib/libconfig.c
index 47920ec..dd41815 100644
--- a/lib/libconfig.c
+++ b/lib/libconfig.c
@@ -1619,7 +1619,12 @@ config_setting_t *config_setting_add(config_setting_t *parent,
   if(! parent)
     return(NULL);
 
-  if((parent->type == CONFIG_TYPE_ARRAY) || (parent->type == CONFIG_TYPE_LIST))
+  if (parrent->type == CONFIG_TYPE_ARRAY &&
+      (type == CONFIG_TYPE_GROUP || type == CONFIG_TYPE_ARRAY ||
+       type == CONFIG_TYPE_LIST))
+    return(NULL);
+
+  if((parent->type == CONFIG_TYPE_ARRAY) || (parent->type == CONFIG_TYPE_LIST)) {
     name = NULL;
 
   if(name)
```

It this is acceptable, please feel free to add it to the code (or I can go through requesting a merge).

**Comment by hyperrealm** [comment]
Thanks for root-causing the problem. I made an equivalent fix in 584a87a01edd60662786274c3fb482824062ccb0
